### PR TITLE
Refactor PIV metadata certificate logic

### PR DESCRIPTION
**Pull request by Joerger**
Changes:
- Sign metadata cert with an ephemeral rsa key instead of the PIV key. This way, touch/pin won't be prompted just to sign and set the metadata certificate.
  - Remove the delayed touch prompt from key generation as it was only needed to sign the metadata cert.
- Instead of returning an error for an unknown certificate in a PIV slot, prompt the user to overwrite the slot.

Once https://github.com/gravitational/teleport/pull/32264 is merged, users will also be prompted to overwrite a slot when the slot's key does not pass the required touch/pin policies.

https://github.com/gravitational/teleport/pull/32334 will fix the lint (goimports) error

### CODE REVIEW

File: api/utils/keys/yubikey.go [hunk 1]
```diff
@@ -56,6 +60,8 @@ var (
 // matching the given touch requirement. This private key will either be newly generated
 // or previously generated by a Teleport client and reused.
 func getOrGenerateYubiKeyPrivateKey(touchRequired bool) (*PrivateKey, error) {
+	ctx := context.TODO()
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:63]: Please don't use context.TODO() unless you are planning a follow up.

In this case, could just add a `ctx` param to the method?

File: api/utils/keys/yubikey.go [hunk 2]
```diff
@@ -70,14 +76,29 @@ func getOrGenerateYubiKeyPrivateKey(touchRequired bool) (*PrivateKey, error) {
...
 
-	// First, check if there is already a private key set up by a Teleport Client.
+	// First, check if there is already a private key on the slot.
 	priv, err := y.getPrivateKey(pivSlot)
-	if trace.IsNotFound(err) {
+	switch {
+	case trace.IsCompareFailed(err):
+		// Found a certificate in the slot that was not generated by a Teleport client.
+		// Prompt the user to confirm before we overwrite the key and cert in this slot.
+		promptQuestion := fmt.Sprintf("%v\nWould you like to overwrite this slot's private key and certificate?", err.Error())
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:85]: nit:

```suggestion
		promptQuestion := fmt.Sprintf("%v\nWould you like to overwrite this slot's private key and certificate?", err)
```

File: api/utils/keys/yubikey.go [hunk 3]
```diff
@@ -144,6 +165,9 @@ func (y *YubiKeyPrivateKey) Public() crypto.PublicKey {
 
 // Sign implements crypto.Signer.
 func (y *YubiKeyPrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	signCtx, cancel := context.WithCancel(context.Background())
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:168]: Similarly, could we add `ctx` to the method signature instead?

File: api/utils/keys/yubikey.go [hunk 4]
```diff
@@ -318,12 +344,17 @@ func (y *yubiKey) getPrivateKey(slot piv.Slot) (*YubiKeyPrivateKey, error) {
 	// Check the slot's certificate to see if it contains a self signed Teleport Client cert.
 	cert, err := yk.Certificate(slot)
 	if err != nil || cert == nil {
-		return nil, trace.NotFound("YubiKey certificate slot is empty, expected a Teleport Client cert")
+		return nil, trace.NotFound("certificate in YubiKey PIV slot %q is empty, expected a Teleport Client cert", slot.String())
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:347]: nit:

```suggestion
		return nil, trace.NotFound("certificate in YubiKey PIV slot %q is empty, expected a Teleport Client cert", slot)
```

File: api/utils/keys/yubikey.go [hunk 5]
```diff
@@ -442,59 +473,78 @@ func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
 	}
 }
 
+// selfSignedMetadataCertificate creates a self signed certificate to be stored in the
+// YubiKey's PIV slot. This certificate is purely used as metadata to determine when a
+// slot is in used by a Teleport Client and is not fit to be used in cryptographic operations.
+func selfSignedMetadataCertificate() (*x509.Certificate, error) {
+	// generate a small rsa key to quickly generate a metadata cert.
+	priv, err := rsa.GenerateKey(rand.Reader, 512)
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:481]: Does it need to be RSA? EC keys are already small and typically faster.

I know this likely doesn't matter, but a key this small makes me uneasy.

File: api/utils/keys/yubikey.go [hunk 6]
```diff
@@ -442,59 +473,78 @@ func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
...
 	return cert, nil
 }
 
-// YubiKeys require touch when generating a private key that requires touch, or using
-// a private key (Sign) with touch required. Unfortunately, there is no good way to
-// check whether touch is cached by the PIV module at a given time. In order to require
-// touch only when needed, we prompt for touch after a short delay when we expect the
-// request would succeed if touch were not required.
+// nonTeleportCertificateError returns a user-readable CompareFailed error.
+// when we run into a non-Teleport certificate, we need to get user input
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:519]:
```suggestion
// When we run into a non-Teleport certificate, we need to get user input
```

File: api/utils/keys/yubikey.go [hunk 7]
```diff
@@ -442,59 +473,78 @@ func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
...
+// before we overwrite it to avoid breaking their non-Teleport PIV programs.
+func nonTeleportCertificateError(slot piv.Slot, cert *x509.Certificate) error {
+	msg := fmt.Sprintf("Certificate in YubiKey PIV slot %q is not a Teleport client cert:\nSlot %s:\n", slot.String(), slot.String())
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:522]: nit:

```suggestion
	msg := fmt.Sprintf("Certificate in YubiKey PIV slot %q is not a Teleport client cert:\nSlot %s:\n", slot, slot)
```

File: api/utils/keys/yubikey.go [hunk 8]
```diff
@@ -442,59 +473,78 @@ func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
...
+	// Gather a small list of user-readable x509 certificate fields to append to the error message.
+	msg += fmt.Sprintf("\tAlgorithm:\t%v\n", cert.SignatureAlgorithm)
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:524]: Please use a strings.Builder or bytes.Buffer. This is too many concats/formatting in a row.

File: api/utils/keys/yubikey_test.go [hunk 9]
```diff
@@ -39,7 +45,8 @@ func TestGetOrGenerateYubiKeyPrivateKey(t *testing.T) {
 	require.NoError(t, err)
 
 	// Test creating a self signed certificate with the key.
-	_, err = selfSignedTeleportClientCertificate(priv, priv.Public())
+	digest := []byte{100}
```
Comment (codingllama) [inline api/utils/keys/yubikey_test.go:48]: This looks rather off, even though I get that it is fake. Are there not size requirements for the digest? I'd expect 32 bytes or something like that.

Eg:

```suggestion
	digest := make([]byte, 32)
```

File: api/utils/keys/yubikey_test.go [hunk 10]
```diff
@@ -53,6 +60,48 @@ func TestGetOrGenerateYubiKeyPrivateKey(t *testing.T) {
...
+	require.NoError(t, err)
+
+	priv, err := rsa.GenerateKey(rand.Reader, 512)
+	require.NoError(t, err)
+	cert, err := metadataCertificateTemplate()
+	require.NoError(t, err)
+	cert.Subject.Organization = []string{"not-teleport"}
+	cert.PublicKey = priv.Public()
+	cert.Raw, err = x509.CreateCertificate(rand.Reader, cert, cert, priv.Public(), priv)
+	require.NoError(t, err)
```
Comment (codingllama) [inline api/utils/keys/yubikey_test.go:89]: Could we do this using production code? Or, even better, test using public APIs?

File: api/utils/keys/yubikey.go (hunk 1 above)
Comment (Joerger) [reply api/utils/keys/yubikey.go:63]: This is followed up by https://github.com/gravitational/teleport/pull/31732

File: api/utils/keys/yubikey.go (hunk 3 above)
Comment (Joerger) [reply api/utils/keys/yubikey.go:168]: If we did, it wouldn't implement `crypto.Signer`. We would need to store a context in the struct which isn't really what we want either. I haven't found any solution like this that I'm happy with though.

File: api/utils/keys/yubikey.go (hunk 1 above)
Comment (codingllama) [reply api/utils/keys/yubikey.go:63]: I would add a comment like `// context.TODO removed by ##31732`, but feel free to ignore it.

File: api/utils/keys/yubikey.go (hunk 3 above)
Comment (codingllama) [reply api/utils/keys/yubikey.go:168]: I would take a comment here too, but otherwise it's alright.

File: api/utils/keys/yubikey_test.go (hunk 9 above)
Comment (Joerger) [reply api/utils/keys/yubikey_test.go:48]: I meant to do `make([]byte, 100)` :facepalm: but I don't think it really matters. I'll use 32.

File: api/utils/keys/yubikey_test.go (hunk 10 above)
Comment (Joerger) [reply api/utils/keys/yubikey_test.go:89]: Can you clarify what you mean by public APIs?

Comment (codingllama) [reply api/utils/keys/yubikey_test.go]: Using public/exported methods.

File: api/utils/keys/yubikey.go [hunk 11]
```diff
@@ -473,62 +475,68 @@ func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
...
+// selfSignedTeleportclientMetadataCertificate creates a self signed certificate to be stored in the
 // YubiKey's PIV slot. This certificate is purely used as metadata to determine when a
 // slot is in used by a Teleport Client and is not fit to be used in cryptographic operations.
-func selfSignedMetadataCertificate() (*x509.Certificate, error) {
-	// generate a small rsa key to quickly generate a metadata cert.
-	priv, err := rsa.GenerateKey(rand.Reader, 512)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
+func selfSignedTeleportclientMetadataCertificate() (*x509.Certificate, error) {
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:484]:
```suggestion
func selfSignedTeleportClientMetadataCertificate() (*x509.Certificate, error) {
```

File: api/utils/keys/yubikey.go [hunk 12]
```diff
@@ -473,62 +475,68 @@ func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
...
-	msg += fmt.Sprintf("\tAlgorithm:\t%v\n", cert.SignatureAlgorithm)
-	msg += fmt.Sprintf("\tSubject DN:\t%v\n", cert.Subject)
-	msg += fmt.Sprintf("\tIssuer DN:\t%v\n", cert.Issuer)
-	msg += fmt.Sprintf("\tSerial:\t\t%v\n", cert.SerialNumber)
 	sum := sha256.Sum256(cert.Raw)
-	msg += fmt.Sprintf("\tFingerprint:\t%v\n", hex.EncodeToString(sum[:]))
-	msg += fmt.Sprintf("\tNot before:\t%v\n", cert.NotBefore)
-	msg += fmt.Sprintf("\tNot after:\t%v\n", cert.NotAfter)
+	fingerPrint := hex.EncodeToString(sum[:])
+	msg := fmt.Sprintf(`Certificate in YubiKey PIV slot %q is not a Teleport client cert:
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:521]: Nicely done.

**Review by codingllama — APPROVED** [review]

File: api/utils/keys/yubikey_test.go (hunk 10 above)
Comment (Joerger) [reply api/utils/keys/yubikey_test.go]: 9c8f581aea128cdc54fcc6018d10a9c4583d4a1d

File: api/utils/keys/yubikey.go [hunk 13]
```diff
@@ -51,10 +51,10 @@ const (
 
 var (
 	// We use slot 9a for Teleport Clients which require `private_key_policy: hardware_key`.
-	pivSlotNoTouch = piv.SlotAuthentication
+	PivSlotNoTouch = piv.SlotAuthentication
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:54]: nit:

```suggestion
	PIVSlotNoTouch = piv.SlotAuthentication
```

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: api/utils/keys/yubikey.go [hunk 14]
```diff
@@ -51,10 +51,10 @@ const (
...
 	// We use slot 9c for Teleport Clients which require `private_key_policy: hardware_key_touch`.
 	// Private keys generated on this slot will use TouchPolicy=Cached.
-	pivSlotWithTouch = piv.SlotSignature
+	PivSlotWithTouch = piv.SlotSignature
```
Comment (codingllama) [inline api/utils/keys/yubikey.go:57]: Ditto.

File: api/utils/keys/yubikey.go [hunk 15]
```diff
@@ -46,38 +51,64 @@ const (
...
 
-	// First, check if there is already a private key set up by a Teleport Client.
+	// First, check if there is already a private key on the slot.
 	priv, err := y.getPrivateKey(pivSlot)
-	if trace.IsNotFound(err) {
+	switch {
+	case trace.IsCompareFailed(err):
+		// Found a certificate in the slot that was not generated by a Teleport client.
+		// Prompt the user to confirm before we overwrite the key and cert in this slot.
+		promptQuestion := fmt.Sprintf("%v\nWould you like to overwrite this slot's private key and certificate?", err)
```
Comment (ibeckermayer) [inline api/utils/keys/yubikey.go:87]: What does printing the `err` here show up as? Is it something we want to display to the user?

If not, we could instead do something like

```go
fmt.Sprintf("Found a certificate in the specified slot that was not generated by a Teleport client.\nWould you like to overwrite this slot's private key and certificate?"
```

File: api/utils/keys/yubikey.go [hunk 16]
```diff
@@ -46,38 +51,64 @@ const (
...
+		if ok, confirmErr := prompt.Confirmation(ctx, os.Stderr, prompt.Stdin(), promptQuestion); confirmErr != nil {
+			return nil, trace.Wrap(confirmErr)
+		} else if !ok {
+			return nil, trace.Wrap(err)
+		}
+
+		// User confirmed overwrite.
+		fallthrough
```
Comment (ibeckermayer) [inline api/utils/keys/yubikey.go:95]: Is `fallthrough` necessary here? If the `trace.IsCompareFailed(err)` case triggers then I don't think we ever fall through to `case trace.IsNotFound(err)`.
```suggestion
```

**Review by ibeckermayer — APPROVED** [review]

File: api/utils/keys/yubikey.go (hunk 15 above)
Comment (Joerger) [reply api/utils/keys/yubikey.go:87]: It is a user-readable error message. In another open PR there is a different compare failed error to display, so I'm going to leave it as is.

File: api/utils/keys/yubikey.go (hunk 16 above)
Comment (Joerger) [reply api/utils/keys/yubikey.go:95]: `fallthrough` will [fall into the next case without checking the case](https://go.dev/play/p/va6R8Oj02z). In this case, we want to generate a new key after a compare failed error, so fallthrough works well here.

**Review by ibeckermayer — APPROVED** [review]

**Review by r0mant — APPROVED** [review]
Bot.

File: api/utils/keys/yubikey.go (hunk 15 above)
Comment (Joerger) [reply api/utils/keys/yubikey.go:87]: FWIW I changed this logic in the other PR so it is more clear where the message is coming from and why db346884e3a0bd69c154ba8a0380d62174283471

**Pull request merged**